### PR TITLE
verify that targetClustersfor UserAccount is dynamically loaded

### PR DESCRIPTION
**Pull request by MatousJobanek**
paired PR https://github.com/codeready-toolchain/host-operator/pull/872

### CODE REVIEW

File: test/e2e/parallel/retarget_test.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+func TestRetargetUserByChangingSpaceTargetCluster(t *testing.T) {
+	// given
+	t.Parallel()
+
+	awaitilities := WaitForDeployments(t)
+	hostAwait := awaitilities.Host()
+	member1Await := awaitilities.Member1()
+	member2Await := awaitilities.Member2()
+
+	userSignup, mur := NewSignupRequest(awaitilities).
```
Comment (alexeykazakov) [inline test/e2e/parallel/retarget_test.go:23]: What do you think about also sharing this space with another user? So the test would cover more complicated case with a space with spacebindgings?

File: test/e2e/usersignup_test.go [hunk 2]
```diff
@@ -197,7 +197,7 @@ func (s *userSignupIntegrationTest) TestProvisionToOtherClusterWhenOneIsFull() {
 			Execute(s.T()).Resources()
 
 		// then
-		require.NotEqual(t, mur1.Spec.UserAccounts[0].TargetCluster, mur2.Spec.UserAccounts[0].TargetCluster)
+		require.NotEqual(t, mur1.Status.UserAccounts[0].Cluster.Name, mur2.Status.UserAccounts[0].Cluster.Name)
```
Comment (mfrancisc) [inline test/e2e/usersignup_test.go:200]: I couldn't find where `mur.Status.UserAccounts.Cluster.Name` is being set in the host-operator and member-operator, I see it's being checked but not set. I'm sure I've missed something. 

But should we use `hostAwait.WaitForSpace(t, mur1.Name, wait.UntilSpaceHasAnyTargetClusterSet())` and same for `mur2` here as well ?

**Review by mfrancisc — COMMENTED** [review]
Nice 👍 

I left only few minor comments/questions.

File: test/e2e/parallel/retarget_test.go (hunk 1 above)
Comment (MatousJobanek) [reply test/e2e/parallel/retarget_test.go:24]: good idea - added another test case ffac8bc0

Comment (MatousJobanek) [reply test/e2e/parallel/retarget_test.go:23]: :man_facepalming:  53353e3d

File: test/e2e/usersignup_test.go (hunk 2 above)
Comment (MatousJobanek) [reply test/e2e/usersignup_test.go:200]: > I couldn't find where mur.Status.UserAccounts.Cluster.Name is being set in the host-operator and member-operator, I see it's being checked but not set. I'm sure I've missed something.

I'm not sure if I follow what you mean. Don't forget that this is a status, so it's set by a MUR controller when it syncs the information with UserAccount https://github.com/codeready-toolchain/host-operator/blob/927edad1f670889d02485c1536a19009bec48e18/controllers/masteruserrecord/sync.go#L95

> But should we use hostAwait.WaitForSpace(t, mur1.Name, wait.UntilSpaceHasAnyTargetClusterSet()) and same for mur2 here as well ?

I probably don't follow your message again. Do you mean that we should have an extra check that would wait until the Space has the needed targetCluster set, etc...? This is already done in the "SignupRequest" fluent API logic - we call it with `EnsureMUR()` which ensures both the MUR and Space are fully provisoned. I know that the name is misleading - it's another thing to improve but in separate PR, I don't want to mix it with the functional changes.

File: testsupport/user_assertions.go [hunk 3]
```diff
@@ -50,29 +73,58 @@ func VerifyUserRelatedResources(t *testing.T, awaitilities wait.Awaitilities, si
...
+		userAccountStatusWaitCriterion)
 	require.NoError(t, err)
 
-	memberAwait := GetMurTargetMember(t, awaitilities, mur)
+	// Verify last target cluster annotation is set
+	lastCluster, foundLastCluster := userSignup.Annotations[toolchainv1alpha1.UserSignupLastTargetClusterAnnotationKey]
+	require.True(t, foundLastCluster)
+	// todo - change that as soon as we set the last-target-cluster annotation based on the "home" Space location
+	// todo - and as soon as we drop the embedded UserAccounts from MUR
+	require.Equal(t, mur.Spec.UserAccounts[0].TargetCluster, lastCluster)
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply testsupport/user_assertions.go:96]: not really - this is one of the things to be modified in the next stories/PRs. The annotation is set when we provision MUR - based on the information that we set directly in the MUR, and not based on what is in Space nor MUR.Status part.

File: testsupport/wait/host.go [hunk 4]
```diff
@@ -471,6 +471,46 @@ func UntilMasterUserRecordHasUserAccountStatuses(expected ...toolchainv1alpha1.U
...
+					if ua.Cluster.Name == expectedCluster {
+						found = true
+						break
+					}
+				}
+				if !found {
+					return false
+				}
+			}
+			return true
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply testsupport/wait/host.go:505]: Content-wise, there is no difference between userAccount1 and userAccount2. They both contain the very same content. The only difference is where they are provisioned.
If a user has access to 100 Spaces and all of them would be provisioned in the same cluster, then there would be only one single UserAccount for the MUR created. If the Spaces would be provisioned to 100 different member clusters, then the MUR would have 100 of UserAccount "replicas", and thus 100 UserAccount items in its status.

The only difference would be if we wanted to propagate the information about the "home" Space to the MUR status as well. In that case, we would need to mark which item in the status corresponds to the cluster where the "home" Space is provisioned. That's all.

File: test/e2e/usersignup_test.go (hunk 2 above)
Comment (mfrancisc) [reply test/e2e/usersignup_test.go:200]: > I'm not sure if I follow what you mean. Don't forget that this is a status, so it's set by a MUR controller when it syncs the information with UserAccount https://github.com/codeready-toolchain/host-operator/blob/927edad1f670889d02485c1536a19009bec48e18/controllers/masteruserrecord/sync.go#L95

Ok I see now that the `mur.Status.UserAccount.Cluster.Name` is being set for the fist time here: I'm not sure if I follow what you mean. Don't forget that this is a status, so it's set by a MUR controller when it syncs the information with UserAccount https://github.com/codeready-toolchain/host-operator/blob/927edad1f670889d02485c1536a19009bec48e18/controllers/masteruserrecord/sync.go#L95

And the UserAccount is being added to the status or updated based on `index` < `0` here:https://github.com/codeready-toolchain/host-operator/blob/927edad1f670889d02485c1536a19009bec48e18/controllers/masteruserrecord/sync.go#L103-L107

Thanks for the pointer!

Comment (mfrancisc) [reply test/e2e/usersignup_test.go:200]: > I probably don't follow your message again. Do you mean that we should have an extra check that would wait until the Space has the needed targetCluster set, etc...? 

Basically I was wondering if we should rely on this `MUR.Status.UserAccounts[0].Cluster.Name` field or if we should get the Spaces for both mur1 and mur1 like we do here for example https://github.com/codeready-toolchain/toolchain-e2e/pull/807/files#diff-d5d12b6ea07b0a78c9ad1987e1486c565ecf16839f77a93e91b09475cf6c0e99R148 and use the `Space.Spec.TargetCluster` for the comparison. But I guess this `MUR.Status.UserAccounts[0].Cluster.Name` will always be up-to-date.

File: testsupport/user_assertions.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mfrancisc) [reply testsupport/user_assertions.go:96]: Got it ! Thanks for explaining that!

File: testsupport/wait/host.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mfrancisc) [reply testsupport/wait/host.go:505]: Makes sense! Thanks for explaining that!

**Review by mfrancisc — APPROVED** [review]
Nice Job 🏅 

Thanks for answering my questions 🙏

File: test/e2e/parallel/retarget_test.go (hunk 1 above)
Comment (mfrancisc) [reply test/e2e/parallel/retarget_test.go:23]: > 🤦‍♂️ https://github.com/codeready-toolchain/toolchain-e2e/commit/53353e3d9d02520cf5ad0120bf3be0205776056b

sometimes I push similar commits 😄 , maybe I'll try to create git pre-commit hook for it.

File: test/e2e/usersignup_test.go (hunk 2 above)
Comment (MatousJobanek) [reply test/e2e/usersignup_test.go:200]: yeah, good idea: done here d424d5ae

Comment (mfrancisc) [reply test/e2e/usersignup_test.go:200]: Nice! Makes sense to have it in addition to the existing check 👍 

Thanks!

**Pull request merged**